src/wallet/coinselection.cpp

```diff
@@ -0,0 +1,349 @@
+// easier to find a solution
+// (now its a simple exponential) const long int LOOP_COST = 500;
+const int P2PKH_INPUT_SIZE = 72; // <Sig> <PubKey> OP_DUP OP_HASH160 <PubkeyHash> OP_EQUALVERIFY OP_CHECKSIG
+
+extern CTweak<unsigned int> maxCoinSelSearchTime;
+extern CTweak<unsigned int> preferredNumUTXO;
+
+// This holds a sorted map of coin amounts and the set of Txos required to get to that amount
+typedef std::multimap<CAmount, TxoGroup> TxoGroupMap;
+
+// How close do we want to get to targetValue (at first).  The longer it takes the more we add to this value.
+CAmount reasonableExcess(CAmount targetValue) { return targetValue / 1024; }
+// Create a group of Txos from a Txo index and another group.
+TxoGroup makeGroup(SpendableTxos::iterator i, const TxoGroup *prev = nullptr)
+{
+    TxoGroup ret;
+    if (prev != nullptr)
+        ret = *prev;
+    else
+        ret.first = 0;
+    // Don't add an element that already exists  -- allow this, bad solutions eliminated later
+    // assert(ret.second.find(i) == ret.second.end());
+    ret.first += i->first; // update the amount
+    COutput outp = i->second;
+    // verify that the output value is consistent -- TODO: Check pcoin and make sure it doesn't get deleted -- located
+    // in mapWallet
+    assert(outp.tx->vout[outp.i].nValue == i->first);
+    ret.second.insert(i); // add this txo to the set
+    return ret;
+}
+
+// Take a group that is not a solution and find a bunch of solutions by appending additional TXOs onto it.
+// This is a recursive function that is limited by MAX_TXOs and MAX_ELECTIVE_TXOS
+bool extendCoinSelectionSolution(const CAmount targetValue,
+    /*const*/ SpendableTxos &available,
+    TxoGroup grp,
+    TxoGroupMap &solutions,
+    int depth)
+{
+    if (depth >= MAX_TXOS)
+        return false; // unambiguously break the recursion
+    SpendableTxos::iterator aEnd = available.end(); // For speed, get these once
+    SpendableTxos::iterator aBeg = available.begin();
+    bool ret = false;
+    SpendableTxos::iterator small;
+    small = available.lower_bound(targetValue - grp.first); // Find a value at or above what we need
+
+    {
+        SpendableTxos::iterator i = small;
+        // while not at the end and there's too little value or we already used this output then keep going.
+        // while((i != aEnd) && ((i->first + grp.first < targetValue) || (grp.second.find(i) != grp.second.end()))  )
+        // allow dups and eliminate later for performance
+        while ((i != aEnd) && (i->first + grp.first < targetValue)) // iterate to make sure it is actually big enough.
+        {
+            ++i;
+        }
+
+        if (i != aEnd) // Ok we have a solution so add it.
+        {
+            TxoGroup g = makeGroup(i, &grp);
+            solutions.insert(TxoGroupMap::value_type(g.first, g));
+            ret = true;
+        }
+    }
+
+    if (small == aEnd)
+        --small;
+    for (int i = 0; i < 5; i++) // check 5 more close values
+    {
```

Comment:
nit:  somewhat confusing when I come back to read the code...above you use i as an iterator then you used it here as an integer...would prefer if you call  the iterator  "it" or "iter" to make the distinction...makes it easier to follow the code.

Reply:
They are used in separate scope which is why we don't get a shadowing warning... I changed the i here to "count"